### ISSUE TITLE
My compilation on caffe fails due to errors in image-data-layer, can you upload the compiled version of caffe?

### ISSUE THREAD

**Issue by iamZe** [wujiyang/Triplet-Loss#2]
Thanks.

**Comment by wujiyang** [comment]
@iamZe  Did you modify the image_data_layer_param in caffe.proto? I added a parameter in it.

**Comment by iamZe** [comment]
你好 我是使用的你的caffe.proto文件直接进行了替换，然后编译的

**Comment by iamZe** [comment]
@wujiyang 请问您方便总结一下在最新版caffe上做哪些变动就可以进行编译了呢，我尝试了好多版本的triplet loss都出错了。

**Comment by wujiyang** [comment]
@iamZe  你可以把image_data_layer的报错问题贴出来看一下，我这些代码直接放到caffe对应的位置就可以编译了。

一般情况下，需要修改的有以下几个方面：
①对应的.h  .cpp .cu文件
②caffe.proto文件中对layer的声明
③申明的layer的相应的参数message
建议你看一下我的修改部分，然后在你的caffe.proto文件中进行修改。

**Comment by iamZe** [comment]
@wujiyang 编译成功了，谢谢~我用过softmax预训练了，为什么起始loss只有0.02呢，这么小很难finetune吧

**Comment by iamZe** [comment]
@wujiyang 这个是我的loss下降记录，学习率设置的0.0001，一直在振荡根本不下降。

    neg_num: 4
    pair_size: 2
    hard_ratio: 0
    rand_ratio: 1
    margin: 0.1

I0703 02:44:16.248906  1599 solver.cpp:218] Iteration 0 (1.35991e-22 iter/s, 1.38166s/100 iters), loss = 0.0270692
I0703 02:44:16.249070  1599 sgd_solver.cpp:105] Iteration 0, lr = 0.0001
I0703 02:46:26.044234  1599 solver.cpp:218] Iteration 100 (0.77045 iter/s, 129.794s/100 iters), loss = 0.0241016
I0703 02:46:26.044363  1599 sgd_solver.cpp:105] Iteration 100, lr = 0.0001
I0703 02:48:35.962218  1599 solver.cpp:218] Iteration 200 (0.769723 iter/s, 129.917s/100 iters), loss = 0.0325377
I0703 02:48:35.962399  1599 sgd_solver.cpp:105] Iteration 200, lr = 0.0001
I0703 02:50:45.770097  1599 solver.cpp:218] Iteration 300 (0.770377 iter/s, 129.807s/100 iters), loss = 0.02679
I0703 02:50:45.770273  1599 sgd_solver.cpp:105] Iteration 300, lr = 0.0001
I0703 02:52:55.635154  1599 solver.cpp:218] Iteration 400 (0.770037 iter/s, 129.864s/100 iters), loss = 0.0261866
I0703 02:52:55.635318  1599 sgd_solver.cpp:105] Iteration 400, lr = 0.0001
I0703 02:55:05.482285  1599 solver.cpp:218] Iteration 500 (0.770143 iter/s, 129.846s/100 iters), loss = 0.032181
I0703 02:55:05.482488  1599 sgd_solver.cpp:105] Iteration 500, lr = 0.0001
I0703 02:57:15.349179  1599 solver.cpp:218] Iteration 600 (0.770026 iter/s, 129.866s/100 iters), loss = 0.0275319
I0703 02:57:15.349342  1599 sgd_solver.cpp:105] Iteration 600, lr = 0.0001
I0703 02:59:25.198025  1599 solver.cpp:218] Iteration 700 (0.770133 iter/s, 129.848s/100 iters), loss = 0.0264444
I0703 02:59:25.198215  1599 sgd_solver.cpp:105] Iteration 700, lr = 0.0001
I0703 03:01:35.043292  1599 solver.cpp:218] Iteration 800 (0.770155 iter/s, 129.844s/100 iters), loss = 0.018997
I0703 03:01:35.043499  1599 sgd_solver.cpp:105] Iteration 800, lr = 0.0001
I0703 03:03:44.975193  1599 solver.cpp:218] Iteration 900 (0.769641 iter/s, 129.931s/100 iters), loss = 0.0327455
I0703 03:03:44.975394  1599 sgd_solver.cpp:105] Iteration 900, lr = 0.0001
I0703 03:05:54.853147  1599 solver.cpp:218] Iteration 1000 (0.769961 iter/s, 129.877s/100 iters), loss = 0.0242539
I0703 03:05:54.853305  1599 sgd_solver.cpp:105] Iteration 1000, lr = 0.0001
^CI0703 03:08:04.760740  1599 solver.cpp:218] Iteration 1100 (0.769785 iter/s, 129.906s/100 iters), loss = 0.0306002

**Comment by wujiyang** [comment]
@iamZe  你的学习率设定的太低了

**Comment by iamZe** [comment]
@wujiyang 尝试过改成0.1,0.01，还是这样不收敛，做了20000轮loop都是这样

**Comment by iamZe** [comment]
I0703 10:54:21.593871 24860 solver.cpp:272] Solving train.prototxt
I0703 10:54:21.593878 24860 solver.cpp:273] Learning Rate Policy: step
I0703 10:54:22.989241 24860 solver.cpp:218] Iteration 0 (0 iter/s, 1.3953s/100 iters), loss = 0.0359277
I0703 10:54:22.989323 24860 sgd_solver.cpp:105] Iteration 0, lr = 0.1
I0703 10:54:28.864961 24860 blocking_queue.cpp:49] Waiting for data
I0703 10:56:34.719353 24860 solver.cpp:218] Iteration 100 (0.759141 iter/s, 131.728s/100 iters), loss = 0.0401807
I0703 10:56:34.720479 24860 sgd_solver.cpp:105] Iteration 100, lr = 0.1
I0703 10:58:46.642441 24860 solver.cpp:218] Iteration 200 (0.758029 iter/s, 131.921s/100 iters), loss = 0.0288354
I0703 10:58:46.643559 24860 sgd_solver.cpp:105] Iteration 200, lr = 0.1
I0703 11:00:58.345880 24860 solver.cpp:218] Iteration 300 (0.759294 iter/s, 131.701s/100 iters), loss = 0.0266514
I0703 11:00:58.347033 24860 sgd_solver.cpp:105] Iteration 300, lr = 0.1
I0703 11:03:10.280844 24860 solver.cpp:218] Iteration 400 (0.757962 iter/s, 131.933s/100 iters), loss = 0.0309502
I0703 11:03:10.281987 24860 sgd_solver.cpp:105] Iteration 400, lr = 0.1

**Comment by wujiyang** [comment]
因为加了norm层，损失本来就很小，我训练的时候迭代2W次左右也没变化多少，最后测试的时候效果还不错。
刚刚测试了一下，当我margin设置为0.2的时候，loss也基本就是从0.2下降到0.1左右，变化范围很小。但是也不至于像你这样这么低。

**Comment by iamZe** [comment]
小也没关系 但是loss并没有在收敛的感觉

**Comment by wujiyang** [comment]
那应该是你自己数据处理的问题了，你再看看吧

**Comment by iamZe** [comment]
好的 谢谢

**Comment by iamZe** [comment]
@wujiyang 最后一个问题。。。请问你在resnet上训练，使用了哪里的预训练model吗，还是自己完全从softmax loss 训练再使用triplet 微调呢？

**Comment by wujiyang** [comment]
我自己从头训练的

**Comment by zhouxiaoxu** [comment]
@wujiyang 你大概训练了多少轮？

**Comment by zhouxiaoxu** [comment]
我训练了几次，每次训练30w次，LFW数据集上的准确率大概89%，感觉很低
